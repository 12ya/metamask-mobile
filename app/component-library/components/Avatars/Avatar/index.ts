export { default } from './Avatar';
export { AvatarSizes, AvatarVariants } from './Avatar.types';
