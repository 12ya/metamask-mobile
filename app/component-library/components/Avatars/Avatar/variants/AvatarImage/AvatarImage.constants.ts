/* eslint-disable import/prefer-default-export */
// Third party dependencies.
import { ImageSourcePropType, ImagePropsBase } from 'react-native';

export const TEST_AVATAR_IMAGE_URI =
  'https://cryptologos.cc/logos/curve-dao-token-crv-logo.png';
/* eslint-disable-next-line */
const TEST_AVATAR_IMAGE_LOCAL_IMAGE_SOURCE: ImageSourcePropType = require('../../../../../../images/fox.png');
const TEST_AVATAR_IMAGE_REMOTE_IMAGE_SOURCE: ImageSourcePropType = {
  uri: TEST_AVATAR_IMAGE_URI,
};
export const TEST_AVATAR_IMAGE_LOCAL_IMAGE_PROPS: ImagePropsBase = {
  source: TEST_AVATAR_IMAGE_LOCAL_IMAGE_SOURCE,
};
export const TEST_AVATAR_IMAGE_REMOTE_IMAGE_PROPS: ImagePropsBase = {
  source: TEST_AVATAR_IMAGE_REMOTE_IMAGE_SOURCE,
};

export const AVATAR_IMAGE_TEST_ID = 'avatar-image';
export const AVATAR_IMAGE_HALO_TEST_ID = 'avatar-image-halo';
export const AVATAR_IMAGE_IMAGE_TEST_ID = 'avatar-image-image';
