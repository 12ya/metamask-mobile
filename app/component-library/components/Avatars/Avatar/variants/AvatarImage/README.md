# AvatarImage

AvatarImage is a component that renders an image contained within an avatar.

## Props

This component extends [AvatarBaseProps](../../foundation/AvatarBase/AvatarBase.types.ts) from the [AvatarBase](../../foundation/AvatarBase/AvatarBase.tsx) component and the ImagePropBase from ReactNative's [Image](https://reactnative.dev/docs/images) component.

### `size`

Optional enum to select between size variants.

| <span style="color:gray;font-size:14px">TYPE</span> | <span style="color:gray;font-size:14px">REQUIRED</span> | <span style="color:gray;font-size:14px">DEFAULT</span> |
| :-------------------------------------------------- | :------------------------------------------------------ | :----------------------------------------------------- |
| [AvatarSizes](../../Avatar.types.ts)          | No                                                     | Md                                                     |

### `backgroundColor`

Optional enum to add color to the background of the Avatar.

| <span style="color:gray;font-size:14px">TYPE</span> | <span style="color:gray;font-size:14px">REQUIRED</span> | <span style="color:gray;font-size:14px">DEFAULT</span> |
| :-------------------------------------------------- | :------------------------------------------------------ | :----------------------------------------------------- |
| [Color](../../../../../constants/typescript.constants.ts) or 'default'          | No                               |'default'                                                     |

### `isHaloEnabled`

Optional boolean to activate halo effect.

| <span style="color:gray;font-size:14px">TYPE</span> | <span style="color:gray;font-size:14px">REQUIRED</span> | <span style="color:gray;font-size:14px">DEFAULT</span> |
| :-------------------------------------------------- | :------------------------------------------------------ | :----------------------------------------------------- |
| boolean          | No                                                     | false                                                     |

### `source`

The image source (either a remote URL or a local file resource). 
The currently supported formats are png, jpg, jpeg, bmp, gif, webp (Android only), psd (iOS only).

| <span style="color:gray;font-size:14px">TYPE</span> | <span style="color:gray;font-size:14px">REQUIRED</span> |
| :-------------------------------------------------- | :------------------------------------------------------ |
| [ImageSourcePropType](https://reactnative.dev/docs/images)                                              | Yes                                                      |

## Usage

```javascript
const imageSource:ImageSourcePropType = {
    uri: SAMPLE_IMAGE_URI
}
<AvatarImage size={AvatarSizes.Md} source={imageSource} isHaloEnabled backgroundColor={'#000000'}/>;
```

