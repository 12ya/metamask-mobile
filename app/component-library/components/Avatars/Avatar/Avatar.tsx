/* eslint-disable react/prop-types */
import React from 'react';

// External dependencies.
import AvatarIcon from './variants/AvatarIcon';
import AvatarImage from './variants/AvatarImage';
import AvatarInitial from './variants/AvatarInitial';
import AvatarJazzIcon from './variants/AvatarJazzIcon';

// Internal dependencies.
import { AvatarProps, AvatarVariants } from './Avatar.types';

const Avatar = (avatarProps: AvatarProps) => {
  switch (avatarProps.variant) {
    case AvatarVariants.Icon:
      return <AvatarIcon {...avatarProps} />;
    case AvatarVariants.Image:
      return <AvatarImage {...avatarProps} />;
    case AvatarVariants.Initial:
      return <AvatarInitial {...avatarProps} />;
    case AvatarVariants.JazzIcon:
      return <AvatarJazzIcon {...avatarProps} />;
    default:
      throw new Error('Invalid Avatar Variant');
  }
};

export default Avatar;
