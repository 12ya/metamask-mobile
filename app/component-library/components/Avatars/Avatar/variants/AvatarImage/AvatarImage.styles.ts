// Third party dependencies.
import { StyleSheet, ImageStyle } from 'react-native';

// External dependencies.
import { Theme } from '../../../../../../util/theme/models';

// Internal dependencies.
import { AvatarImageStyleSheetVars } from './AvatarImage.types';

/**
 * Style sheet function for AvatarImage component.
 *
 * @param params Style sheet params.
 * @param params.theme App theme from ThemeContext.
 * @param params.vars Inputs that the style sheet depends on.
 * @returns StyleSheet object.
 */
const styleSheet = (params: {
  theme: Theme;
  vars: AvatarImageStyleSheetVars;
}) => {
  const { vars } = params;
  const { size, isHaloEnabled } = vars;
  const imageSize = isHaloEnabled ? Number(size) * 0.64 : Number(size);
  return StyleSheet.create({
    haloImage: {
      opacity: 0.5,
    },
    halo: {
      width: Number(size),
      height: Number(size),
      justifyContent: 'center',
      alignItems: 'center',
      overflow: 'hidden',
      borderRadius: Number(size) / 2,
    },
    image: Object.assign({
      width: imageSize,
      height: imageSize,
      borderRadius: imageSize / 2,
    } as ImageStyle),
  });
};

export default styleSheet;
