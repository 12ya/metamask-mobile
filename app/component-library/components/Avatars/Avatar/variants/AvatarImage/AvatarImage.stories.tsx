// Third party dependencies.
import React from 'react';
import { select, boolean } from '@storybook/addon-knobs';

// External dependencies.
import { storybookPropsGroupID } from '../../../../../constants/storybook.constants';
import { AvatarSizes } from '../../Avatar.types';
import {
  BadgeProps,
  BadgeVariants,
} from '../../../../Badges/Badge/Badge.types';
import { TEST_AVATAR_PROPS } from '../../../../Badges/Badge/variants/BadgeAvatar/BadgeAvatar.constants';

// Internal dependencies.
import AvatarImage from './AvatarImage';
import { AvatarImageProps } from './AvatarImage.types';
import { TEST_AVATAR_IMAGE_REMOTE_IMAGE_SOURCE } from './AvatarImage.constants';

export const getAvatarImageStoryProps = (): AvatarImageProps => {
  const sizeSelector = select(
    'size',
    AvatarSizes,
    AvatarSizes.Md,
    storybookPropsGroupID,
  );

  const isHaloEnabled = boolean('isHaloEnabled', false, storybookPropsGroupID);

  const includeBadge = boolean('includeBadge', false, storybookPropsGroupID);

  const badgeProps: BadgeProps = {
    variant: BadgeVariants.Avatar,
    avatarProps: TEST_AVATAR_PROPS,
  };

  return {
    size: sizeSelector,
    source: TEST_AVATAR_IMAGE_REMOTE_IMAGE_SOURCE,
    isHaloEnabled,
    includeBadge,
    badgeProps,
  };
};
const AvatarImageStory = () => <AvatarImage {...getAvatarImageStoryProps()} />;

export default AvatarImageStory;
