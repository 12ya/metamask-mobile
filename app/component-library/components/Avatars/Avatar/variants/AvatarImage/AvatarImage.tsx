// Third party dependencies.
import React from 'react';
import { Image, ImageBackground } from 'react-native';

// External dependencies.
import { useStyles } from '../../../../../hooks';
import AvatarBase from '../../foundation/AvatarBase';
import { DEFAULT_AVATAR_SIZE } from '../../Avatar.constants';

// Internal dependencies.
import styleSheet from './AvatarImage.styles';
import { AvatarImageProps } from './AvatarImage.types';
import {
  AVATAR_IMAGE_TEST_ID,
  AVATAR_IMAGE_IMAGE_TEST_ID,
} from './AvatarImage.constants';

const AvatarImage = ({
  size = DEFAULT_AVATAR_SIZE,
  imageProps,
  ...props
}: AvatarImageProps) => {
  const { styles } = useStyles(styleSheet, { size });

  return (
    <AvatarBase size={size} testID={AVATAR_IMAGE_TEST_ID} {...props}>
      <Image
        style={styles.image}
        resizeMode={'contain'}
        testID={AVATAR_IMAGE_IMAGE_TEST_ID}
        {...imageProps}
      />
    </AvatarBase>
  );
};

export default AvatarImage;
