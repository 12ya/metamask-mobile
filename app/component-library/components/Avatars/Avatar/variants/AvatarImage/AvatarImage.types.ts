// Third party dependencies.
import { ImagePropsBase } from 'react-native';

// External dependencies.
import { AvatarBaseProps } from '../../foundation/AvatarBase';
import { AvatarVariants } from '../../Avatar.types';

/**
 * AvatarImage component props.
 */
export type AvatarImageProps = AvatarBaseProps &
  ImagePropsBase & {
    /**
     * Avatar variants.
     */
    variant?: AvatarVariants.Image;
    /**
     * Optional boolean to activate halo effect.
     * @default false
     */
    isHaloEnabled?: boolean;
  };

/**
 * Style sheet input parameters.
 */
export type AvatarImageStyleSheetVars = Pick<
  AvatarImageProps,
  'size' | 'style' | 'isHaloEnabled'
>;
