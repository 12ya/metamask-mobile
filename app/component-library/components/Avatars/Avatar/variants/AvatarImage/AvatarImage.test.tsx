// Third party dependencies.
import React from 'react';
import { shallow } from 'enzyme';

// External dependencies.
import { AvatarSize, AvatarVariants } from '../../Avatar.types';

// Internal dependencies.
import AvatarImage from './AvatarImage';
import {
  TEST_AVATAR_IMAGE_LOCAL_IMAGE_SOURCE,
  TEST_AVATAR_IMAGE_REMOTE_IMAGE_SOURCE,
  AVATAR_IMAGE_TEST_ID,
  AVATAR_IMAGE_HALO_TEST_ID,
  AVATAR_IMAGE_IMAGE_TEST_ID,
} from './AvatarImage.constants';

describe('AvatarImage - Snapshot', () => {
  it('should render AvatarImage with local imageSource correctly', () => {
    const wrapper = shallow(
      <AvatarImage
        variant={AvatarVariants.Image}
        size={AvatarSize.Md}
        imageSource={TEST_AVATAR_IMAGE_LOCAL_IMAGE_SOURCE}
      />,
    );
    expect(wrapper).toMatchSnapshot();
  });
  it('should render AvatarImage with remote imageSource correctly', () => {
    const wrapper = shallow(
      <AvatarImage
        variant={AvatarVariants.Image}
        size={AvatarSize.Md}
        imageSource={TEST_AVATAR_IMAGE_REMOTE_IMAGE_SOURCE}
      />,
    );
    expect(wrapper).toMatchSnapshot();
  });
  it('should render AvatarImage with halo if isHaloEnabled=true', () => {
    const wrapper = shallow(
      <AvatarImage
        variant={AvatarVariants.Image}
        size={AvatarSize.Md}
        imageSource={TEST_AVATAR_IMAGE_LOCAL_IMAGE_SOURCE}
        isHaloEnabled
      />,
    );
    expect(wrapper).toMatchSnapshot();
  });
});

describe('AvatarImage', () => {
  it('should render AvatarImage component', () => {
    const wrapper = shallow(
      <AvatarImage
        variant={AvatarVariants.Image}
        size={AvatarSize.Md}
        imageSource={TEST_AVATAR_IMAGE_REMOTE_IMAGE_SOURCE}
      />,
    );
    const AvatarImageComponent = wrapper.findWhere(
      (node) => node.prop('testID') === AVATAR_IMAGE_TEST_ID,
    );
    expect(AvatarImageComponent.exists()).toBe(true);
  });
  it('should render AvatarImage component with local imageSource', () => {
    const wrapper = shallow(
      <AvatarImage
        variant={AvatarVariants.Image}
        size={AvatarSize.Md}
        imageSource={TEST_AVATAR_IMAGE_LOCAL_IMAGE_SOURCE}
      />,
    );
    const AvatarImageImageComponent = wrapper.findWhere(
      (node) => node.prop('testID') === AVATAR_IMAGE_IMAGE_TEST_ID,
    );
    expect(AvatarImageImageComponent.props().source).toBe(
      TEST_AVATAR_IMAGE_LOCAL_IMAGE_SOURCE,
    );
  });
  it('should render AvatarImage component with remote imageSource', () => {
    const wrapper = shallow(
      <AvatarImage
        variant={AvatarVariants.Image}
        size={AvatarSize.Md}
        imageSource={TEST_AVATAR_IMAGE_REMOTE_IMAGE_SOURCE}
      />,
    );
    const AvatarImageImageComponent = wrapper.findWhere(
      (node) => node.prop('testID') === AVATAR_IMAGE_IMAGE_TEST_ID,
    );
    expect(AvatarImageImageComponent.props().source).toBe(
      TEST_AVATAR_IMAGE_REMOTE_IMAGE_SOURCE,
    );
  });
  it('should render AvatarImage with halo if isHaloEnabled=true', () => {
    const wrapper = shallow(
      <AvatarImage
        variant={AvatarVariants.Image}
        size={AvatarSize.Md}
        imageSource={TEST_AVATAR_IMAGE_REMOTE_IMAGE_SOURCE}
        isHaloEnabled
      />,
    );
    const AvatarImageHaloComponent = wrapper.findWhere(
      (node) => node.prop('testID') === AVATAR_IMAGE_HALO_TEST_ID,
    );
    expect(AvatarImageHaloComponent.exists()).toBe(true);
  });
  it('should NOT render AvatarImage with halo if isHaloEnabled=false', () => {
    const wrapper = shallow(
      <AvatarImage
        variant={AvatarVariants.Image}
        size={AvatarSize.Md}
        imageSource={TEST_AVATAR_IMAGE_REMOTE_IMAGE_SOURCE}
        isHaloEnabled={false}
      />,
    );
    const AvatarImageHaloComponent = wrapper.findWhere(
      (node) => node.prop('testID') === AVATAR_IMAGE_HALO_TEST_ID,
    );
    expect(AvatarImageHaloComponent.exists()).toBe(false);
  });
});
