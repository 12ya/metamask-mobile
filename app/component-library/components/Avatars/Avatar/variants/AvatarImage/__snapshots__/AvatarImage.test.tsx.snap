// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AvatarImage - Snapshot should render AvatarImage with halo if isHaloEnabled=true 1`] = `
<AvatarBase
  size="32"
  style={Object {}}
  testID="avatar-image"
>
  <ImageBackground
    blurRadius={20}
    imageStyle={
      Object {
        "opacity": 0.5,
      }
    }
    source={
      Object {
        "testUri": "../../../app/images/fox.png",
      }
    }
    style={
      Object {
        "alignItems": "center",
        "borderRadius": 16,
        "height": 32,
        "justifyContent": "center",
        "overflow": "hidden",
        "width": 32,
      }
    }
    testID="avatar-image-halo"
  >
    <Image
      resizeMode="contain"
      source={
        Object {
          "testUri": "../../../app/images/fox.png",
        }
      }
      style={
        Object {
          "borderRadius": 10.24,
          "height": 20.48,
          "width": 20.48,
        }
      }
      testID="avatar-image-image"
      variant="Image"
    />
  </ImageBackground>
</AvatarBase>
`;

exports[`AvatarImage - Snapshot should render AvatarImage with local imageSource correctly 1`] = `
<AvatarBase
  size="32"
  style={Object {}}
  testID="avatar-image"
>
  <Image
    resizeMode="contain"
    source={
      Object {
        "testUri": "../../../app/images/fox.png",
      }
    }
    style={
      Object {
        "borderRadius": 16,
        "height": 32,
        "width": 32,
      }
    }
    testID="avatar-image-image"
    variant="Image"
  />
</AvatarBase>
`;

exports[`AvatarImage - Snapshot should render AvatarImage with remote imageSource correctly 1`] = `
<AvatarBase
  size="32"
  style={Object {}}
  testID="avatar-image"
>
  <Image
    resizeMode="contain"
    source={
      Object {
        "uri": "https://cryptologos.cc/logos/curve-dao-token-crv-logo.png",
      }
    }
    style={
      Object {
        "borderRadius": 16,
        "height": 32,
        "width": 32,
      }
    }
    testID="avatar-image-image"
    variant="Image"
  />
</AvatarBase>
`;
